Guilherme Kenzo, Gustavo Hess

Part 2 - Heteroskesdacity

Question 18: Explain the problem of heteroskedasticity with an example of the course.




Heteroskedasticity happens when the variance across the errors in a regression model are not constant, they vary acording to value of the independent variables.

A very simple way to visualize this concept is with the following example: imagine we plot the age and weight of a lot of people (age x weight). In the beggining of the graph we'll find a very low variance in the weights, considering that althought there are weight differences between babies and children, this difference is hardly too dramatic. Now, as age grows, we'll see a fraction of people gaining a lot of weight in comparisson to the average, we'll also see another fraction losing weight in comparisson to the average. This will certainly increase variance in comparisson with the variance in the first years of human development.

Now we can see an example given:

$$
\sigma_i^2 = x_{i1}^2 + x_{i2}^2 + \dots + x_{iK}^2
$$
Here, the variance of errors depends on the squares of the independent variables. This means the dispertion of errors will increase with the independent variables

The full model specification becomes:

$$
y_i = \beta_0 + \sum_{k=1}^{K} \beta_k x_{ik} + u_i
$$

With the heteroskedasticity assumption:

$$
\text{Var}(u_i | x_{i1}, x_{i2}, \dots, x_{iK}) = x_{i1}^2 + x_{i2}^2 + \dots + x_{iK}^2
$$



Question 19: In the specification of question 9, test the hypothesis of no heteroskedasticity of linear form, i.e. in the regression of u2 on constant, crime, nox, rooms, proptax, test H0: crime, nox, room, proptax = 0, where the coefficients k (k = crime, nox, rooms, proptax) are associated with the corresponding explanatory variables.

In [2]:
import pandas as pd

df = pd.read_csv('HPRICE2.RAW', delim_whitespace=True, decimal='.', header=None)
df = df.apply(pd.to_numeric) 

df = df.rename(columns={
    0: "price",
    1: "crime",
    2: "nox",
    3: "rooms",
    4: "dist",
    5: "radial",
    6: "proptax",
    7: "stratio",
    8: "lowstat",
    9: "lprice",
    10: "lnox",
    11: "lproptax",
})

C:\Users\guik2\AppData\Local\Temp\ipykernel_22820\657320903.py:3: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  df = pd.read_csv('HPRICE2.RAW', delim_whitespace=True, decimal='.', header=None)


In [3]:
import statsmodels.api as sm

# this is the dataframe with the independent variables
X = df[['crime', 'nox', 'rooms', 'proptax']]
X = sm.add_constant(X)

# this is the dataframe with the dependent variable
y = df['price']

# running the regression
model = sm.OLS(y, X).fit()

# now, making the regression of u2 on the appointed variables
u=model.resid
u2=u**2
y=u2
model=sm.OLS(y,X)
results = model.fit()

print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.051
Model:                            OLS   Adj. R-squared:                  0.044
Method:                 Least Squares   F-statistic:                     6.799
Date:                Sun, 08 Dec 2024   Prob (F-statistic):           2.47e-05
Time:                        14:47:50   Log-Likelihood:                -10130.
No. Observations:                 506   AIC:                         2.027e+04
Df Residuals:                     501   BIC:                         2.029e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -5.885e+07   6.41e+07     -0.917      0.3

Only 5.1% of the error variation can be explained by the independent variables (R-Squared).
Probability of F-Statistic is very low, which means the chances of gettig these results given homoskedascity are very low, meaning we are probably looking at heteroskesdacity. We don't really know at this point which variable is causing this, since the F-Statistic looks at the collective impact of the variables.
And the p-value for all our variables are pretty high except for proptax. This means proptax is significately associated with the residues variance: we reject homoskesdacity hypothesis.



Question 20: In the specification of question 10, test the hypothesis of no heteroskedasticity of linear form.

In [ ]:
import statsmodels.api as sm

# this is the dataframe with the independent variables
X = df[['crime', 'nox', 'rooms', 'proptax']]
X = sm.add_constant(X)

# this is the dataframe with the dependent variable
y = df['lprice']

# running the regression
model = sm.OLS(y, X).fit()

# now, making the regression of u2 on the appointed variables
u=model.resid
u2=u**2
y=u2
model=sm.OLS(y,X)
results = model.fit()

from scipy.stats import f

alpha = 0.05
df1 = results.df_model 
df2 = results.df_resid

f_critical = f.ppf(1 - alpha, df1, df2)

print(results.summary())
print("F-critical: ",f_critical)

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.126
Model:                            OLS   Adj. R-squared:                  0.119
Method:                 Least Squares   F-statistic:                     17.98
Date:                Sun, 08 Dec 2024   Prob (F-statistic):           8.33e-14
Time:                        17:43:21   Log-Likelihood:                 185.10
No. Observations:                 506   AIC:                            -360.2
Df Residuals:                     501   BIC:                            -339.1
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0757      0.090      0.842      0.4

By assuming alpha 5% -- we found F-statistic > F-critical: so we reject the null hypothesis -- at least one of the independent variables explains the change in the dependent variable. This points to heteroskesdacity.
And now, we can see that not only proptax p-value is small, the rooms p-value is also small. That means these two independent variables are explaining the change in the error.

Question 21: In the specification of question 11, test the hypothesis of no heteroskedasticity of linear form.

In [14]:
import statsmodels.api as sm

# this is the dataframe with the independent variables
X = df[['crime', 'lnox', 'rooms', 'lproptax']]
X = sm.add_constant(X)

# this is the dataframe with the dependent variable
y = df['lprice']

# running the regression
model = sm.OLS(y, X).fit()

# now, making the regression of u2 on the appointed variables
u=model.resid
u2=u**2
y=u2
model=sm.OLS(y,X)
results = model.fit()

print(results.summary())

alpha = 0.05
df1 = results.df_model   
df2 = results.df_resid   

f_critical = f.ppf(1 - alpha, df1, df2)

print("F-critical: ",f_critical)

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.116
Model:                            OLS   Adj. R-squared:                  0.109
Method:                 Least Squares   F-statistic:                     16.51
Date:                Sun, 08 Dec 2024   Prob (F-statistic):           1.02e-12
Time:                        17:48:15   Log-Likelihood:                 184.66
No. Observations:                 506   AIC:                            -359.3
Df Residuals:                     501   BIC:                            -338.2
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.5496      0.167     -3.289      0.0

By assuming alpha 5% -- we found F-statistic > F-critical: so we reject the null hypothesis -- at least one of the independent variables explains the change in the dependent variable. This points to heteroskesdacity.

We can also see that rooms in marginaly statisticaly significant (again, assuming 5%) and lproptax is higly significant in explaining the error. This means these two variables are explaining the variance in the error and thus are responsible for the high F-Statistic.



Question 22: Comment on the differences between your results of questions 20 and 21.

- The question 20 refers to the model built in question 10 which uses lprice as the dependent variable, and constant, crime, nox, rooms, proptax as the independent variables. The model showed reasonably good fit R-squared: 0.611 and the p-value of all the coeficients are very low, meaning they are all statistically significant. Now, about the results of question 20: we found a big value for the F-statistic, meaning our independent variables can collectively explain the change in error, this point to heteroskesdacity. And looking at the coeficients we can see that the p-values of proptax and rooms are small, meaning they are the variables explaining the change and making the F-statistic high.

- Now, for question 21 we utilized the model built on question 11 where the only difference in comparrison to question 10 is that we utilized lnox instead of nox and lproptax instead of proptax. In these results, we also got a high value for the F-statistic and uppon closer inspection, we can see that lproptax is still explaining the error change and rooms is now marginaly statisticaly significant (close to 5%).

Question 23: Using the specification of question 9, identify the most significant variable causing heteroskedasticity using the student statistics and run a WLS regression with the identified variable as weight. Compare the standards errors with those of question 9. Comment on your results.

This is just the regression made in the question 9:

In [19]:
import statsmodels.api as sm

X = df[['crime', 'nox', 'rooms', 'proptax']]
X = sm.add_constant(X)

y = df['price']

model = sm.OLS(y, X).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.575
Model:                            OLS   Adj. R-squared:                  0.572
Method:                 Least Squares   F-statistic:                     169.6
Date:                Sun, 08 Dec 2024   Prob (F-statistic):           1.06e-91
Time:                        18:32:34   Log-Likelihood:                -5119.6
No. Observations:                 506   AIC:                         1.025e+04
Df Residuals:                     501   BIC:                         1.027e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -1.868e+04   3215.205     -5.811      0.0

This is the error regression to test for heteroskesdacity:

In [18]:
# redisuals
u = model.resid
u2 = u**2

# testing for heteroskedasticity
Z = df[['crime', 'nox', 'rooms', 'proptax']]
Z = sm.add_constant(Z)
model_u = sm.OLS(u2, Z).fit()
print(model_u.summary())


                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.051
Model:                            OLS   Adj. R-squared:                  0.044
Method:                 Least Squares   F-statistic:                     6.799
Date:                Sun, 08 Dec 2024   Prob (F-statistic):           2.47e-05
Time:                        18:32:23   Log-Likelihood:                -10130.
No. Observations:                 506   AIC:                         2.027e+04
Df Residuals:                     501   BIC:                         2.029e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -5.885e+07   6.41e+07     -0.917      0.3

We can see here that the p-value of the proptax is the only low one -- this is the variable explaining the variation in the error and causing heteroskedacity. This is the variable we got to chose to use as weight in the WLS model:

In [35]:
W = 1 / df['proptax']

X = df[['crime', 'nox', 'rooms', 'proptax']]
X = sm.add_constant(X)
y = df['price']

wls_model = sm.WLS(y, X, weights=W).fit()
print(wls_model.summary())


                            WLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.675
Model:                            WLS   Adj. R-squared:                  0.673
Method:                 Least Squares   F-statistic:                     260.6
Date:                Sun, 08 Dec 2024   Prob (F-statistic):          6.53e-121
Time:                        18:52:23   Log-Likelihood:                -5063.3
No. Observations:                 506   AIC:                         1.014e+04
Df Residuals:                     501   BIC:                         1.016e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       -2.99e+04   2820.558    -10.601      0.0

In [32]:
# let's analyze changes in std error:
model_standard_errors = model.bse
wls_model_standard_errors = wls_model.bse

print("For the initial model: \n")
print(model_standard_errors, "\n")
print("For the WLS regression: \n")
print(wls_model_standard_errors)

For the initial model: 

const      3215.204520
crime        38.528570
nox         314.633813
rooms       404.345434
proptax      23.965587
dtype: float64 

For the WLS regression: 

const      2820.558498
crime        44.804955
nox         279.170215
rooms       348.497559
proptax      23.380653
dtype: float64


We can see that the std errors mostly lower for the WLS regression, meaning the method actually worked for dealing with heteroskedasticity. Another thing we can see is that the R-squared also got better: we had 0.575 and now we got 0.675 pointing that our model does a better job at explaining the dependent variable changes.